source/common/protobuf/utility.cc

```diff
@@ -168,21 +168,21 @@ void tryWithApiBoosting(MessageXformFn f, Protobuf::Message& message) {
   Protobuf::DynamicMessageFactory dmf;
   auto earlier_message = ProtobufTypes::MessagePtr(dmf.GetPrototype(earlier_version_desc)->New());
   ASSERT(earlier_message != nullptr);
-  try {
+  TRY_NEEDS_AUDIT {
```

Comment:
Should be main thread only

Reply:
This is called by loadFromJson and actually gets executed in worker thread in a test.

Reply:
Which test? I would be surprised if this happens on workers.

Reply:
RUN      ] IpVersions/SdsStaticDownstreamIntegrationTest.RouterRequestAndResponseWithGiantBodyBuffer/IPv4
[2021-03-04 20:33:07.657][15][critical][assert] [source/common/protobuf/utility.cc:172] assert failure: Thread::MainThread::isMainThread().
[2021-03-04 20:33:07.657][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Aborted, suspect faulting address 0xfffe0000000f
[2021-03-04 20:33:07.657][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2021-03-04 20:33:07.657][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: 0/1.18.0-dev/test/DEBUG/BoringSSL
[2021-03-04 20:33:07.703][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x47b482c]
[2021-03-04 20:33:07.703][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: __restore_rt [0x7ff1081cf980]
[2021-03-04 20:33:07.750][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: Envoy::MessageUtil::loadFromJson() [0x4a52807]
[2021-03-04 20:33:07.795][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: Envoy::(anonymous namespace)::jsonConvertInternal() [0x4a54339]
[2021-03-04 20:33:07.839][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: Envoy::MessageUtil::loadFromYaml() [0x4a53a95]
[2021-03-04 20:33:07.884][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: Envoy::TestUtility::loadFromYaml() [0x28c2822]
[2021-03-04 20:33:07.928][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Ssl::createClientSslTransportSocketFactory() [0x28c142f]
[2021-03-04 20:33:07.972][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: Envoy::Ssl::SdsStaticDownstreamIntegrationTest::initialize() [0x27af8ff]
[2021-03-04 20:33:08.017][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: Envoy::HttpIntegrationTest::testRouterRequestAndResponseWithBody() [0x281b320]
[2021-03-04 20:33:08.061][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Ssl::SdsStaticDownstreamIntegrationTest_RouterRequestAndResponseWithGiantBodyBuffer_Test::TestBody() [0x27abff7]
[2021-03-04 20:33:08.106][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x5b6a884]
[2021-03-04 20:33:08.150][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #11: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x5b59afb]
[2021-03-04 20:33:08.195][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #12: testing::Test::Run() [0x5b458e3]
[2021-03-04 20:33:08.240][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #13: testing::TestInfo::Run() [0x5b462b7]
[2021-03-04 20:33:08.285][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #14: testing::TestSuite::Run() [0x5b46a6c]
[2021-03-04 20:33:08.330][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #15: testing::internal::UnitTestImpl::RunAllTests() [0x5b52a0f]
[2021-03-04 20:33:08.375][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #16: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x5b6e654]
[2021-03-04 20:33:08.420][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #17: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x5b5bd2b]
[2021-03-04 20:33:08.465][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #18: testing::UnitTest::Run() [0x5b525b3]
[2021-03-04 20:33:08.510][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #19: RUN_ALL_TESTS() [0x43c6fc1]
[2021-03-04 20:33:08.554][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #20: Envoy::TestRunner::RunTests() [0x43c6800]
[2021-03-04 20:33:08.599][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #21: main [0x43c4a31]
[2021-03-04 20:33:08.599][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #22: __libc_start_main [0x7ff107dedbf7]

Reply:
Envoy::Ssl::createClientSslTransportSocketFactory()
can happen on worker thread.

Reply:
[ RUN      ] IpVersions/SdsStaticDownstreamIntegrationTest.RouterRequestAndResponseWithGiantBodyBuffer/IPv4
[2021-03-04 20:33:07.657][15][critical][assert] [source/common/protobuf/utility.cc:172] assert failure: Thread::MainThread::isMainThread().
[2021-03-04 20:33:07.657][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Aborted, suspect faulting address 0xfffe0000000f
[2021-03-04 20:33:07.657][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2021-03-04 20:33:07.657][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: 0/1.18.0-dev/test/DEBUG/BoringSSL
[2021-03-04 20:33:07.703][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x47b482c]
[2021-03-04 20:33:07.703][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: __restore_rt [0x7ff1081cf980]
[2021-03-04 20:33:07.750][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: Envoy::MessageUtil::loadFromJson() [0x4a52807]
[2021-03-04 20:33:07.795][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: Envoy::(anonymous namespace)::jsonConvertInternal() [0x4a54339]
[2021-03-04 20:33:07.839][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: Envoy::MessageUtil::loadFromYaml() [0x4a53a95]
[2021-03-04 20:33:07.884][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: Envoy::TestUtility::loadFromYaml() [0x28c2822]
[2021-03-04 20:33:07.928][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Ssl::createClientSslTransportSocketFactory() [0x28c142f]
[2021-03-04 20:33:07.972][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: Envoy::Ssl::SdsStaticDownstreamIntegrationTest::initialize() [0x27af8ff]
[2021-03-04 20:33:08.017][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: Envoy::HttpIntegrationTest::testRouterRequestAndResponseWithBody() [0x281b320]
[2021-03-04 20:33:08.061][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Ssl::SdsStaticDownstreamIntegrationTest_RouterRequestAndResponseWithGiantBodyBuffer_Test::TestBody() [0x27abff7]
[2021-03-04 20:33:08.106][15][critical][backtrace] [bazel-out/k8-fastbuild/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: testing::internal::HandleSehExceptionsInMethodIfSupported<>() [0x5b6a884]

Reply:
I believe the factory is only created on the main thread.

Reply:
This looks like it's being used on the main test thread rather than Envoy main server thread.

Reply:
Can the main test thread be different from main server thread?

Reply:
Yes, I test it. The main test thread is different from the main server thread.

Reply:
We can fix this by registering all the work threads id in a set and change the assertion to NotInWorkerthread.

Reply:
The test failure is actually caused by the different thread id of test main thread and server main thread, I have fixed this problem by registering both the test main thread is and server main thread id in main thread singleton in the integration tests.